File: Smart_labs/simpleAlg.hs
{-
    Basic algebra algs
-}

thirdDouble (x, y, z) = (x, y, z*2)

twoDegree :: Int -> Int -> (Int, Int, Int)
twoDegree x y
    | mod x 2 == 0 && mod y 2 == 0 = thirdDouble $ twoDegree (x `div` 2) (y `div` 2)  
    | otherwise = (x, y, 1)

euclidStep :: Int -> Int -> Int
euclidStep x y
    | x == 1 || y == 1 = 1
    | x == y = x
    | mod x 2 == 0 = euclidStep (div x 2) y
    | mod y 2 == 0 = euclidStep x (div y 2)
    | otherwise = euclidStep (abs(x-y) `div` 2) $ min x y

--binary euclid algoritm
euclid :: Int -> Int -> Int
euclid a b = (trd $ twoDegInfo) * (euclidStep (fstM twoDegInfo) (sndM twoDegInfo) )
    where
        fstM (x, y, z) = x
        sndM (x, y, z) = y
        trd (x, y, z) = z        
        twoDegInfo = twoDegree a b
        
        

File: Smart_labs/simpleAlg.py
"""
Basic algebra algs
"""

def partitions(iter1, iter2):
    return [(x, y) for x in iter1 for y in iter2]

def euclid_GCD(a, b):
    def euclid_step(x, y):
        if x == 1 or y == 1:
            return 1
        if x == y:
            return x
        if x % 2 == 0:
            return euclid_step(x // 2, y)
        if y % 2 == 0:
            return euclid_step(x, y // 2)
        return euclid_step(abs(x - y) // 2, min(x, y))

    gcd = 1
    while a % 2 == 0 and b % 2 == 0:
        a //= 2
        b //= 2
        gcd *= 2

    return gcd * euclid_step(a, b)


